### PR TITLE
Try fast_reject::simplify_type in coherence before doing full check

### PR DESCRIPTION
**Pull request by rylev**
This is a reattempt at landing #69010 (by @jonas-schievink). The change adds a fast path for coherence checking to see if there's no way for types to unify since full coherence checking can be somewhat expensive.

This has big effects on code generated by the [`windows`](https://github.com/microsoft/windows-rs) which in some cases spends as much as 20% of compilation time in the `specialization_graph_of` query. In local benchmarks this took a compilation that previously took ~500 seconds down to ~380 seconds.

This is surely not going to make a difference on much smaller crates, so the question is whether it will have a negative impact. #69010 was closed because some of the perf suite crates did show small regressions.   

Additional discussion of this issue is happening [here](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/windows-rs.20perf).

### CODE REVIEW

**Comment by rylev** [comment]
The perf regression looks better here than in #69010 - only impacting packed-simd. The only regression over 1% is packed-simd-doc which has a [2.4% increase](https://perf.rust-lang.org/detailed-query.html?commit=314cb49364173f1ac3a9566663aabfd4d22446c5&base_commit=e708cbd91c9cae4426d69270248362b423324556&benchmark=packed-simd-doc&run_name=full) in the `specialization_graph_of` query. Basically all other benchmarks besides the various packed-simd ones are unaffected or are small perf gains.

**Comment by Mark-Simulacrum** [comment]
IMO the regressions in this PR are entirely acceptable given the (unfortunately not yet tracked) huge wins we're seeing on some crates, such as windows-rs.

**Comment by rylev** [comment]
I did some benchmarking on the packed-simd case. It seems there that the optimization successfully short-circuits the larger coherence check ~13% of the time. Coherence checking is a larger percentage of time in the docs case which is why this is extra pronounced in the docs compilation. The windows-rs benchmark triggers the shortcutting over 99% of the time so it's much more useful. 

This makes sense as in the worst case scenario this change will *only* add to the time it takes to do coherence checking. The question is how often this shortcut will trigger in the average case.

**Review by lcnr — COMMENTED** [review]
this does not reject `impl Trait for &u32` and `impl Trait for u32` because `simplify_type` was originally intended to be used in typeck where we look into references.

Does the hitrate go up if you extend `simplify_type` to optionally emit another variant for references?

**Comment by rylev** [comment]
@lcnr this does indeed lead to more rejections but it seems to slow things down elsewhere in the `specialization_graph_of` query. I need to figure out what's slowing things down.

**Comment by rylev** [comment]
Queuing another performance run.

@bors try @rust-timer queue

**Comment by rylev** [comment]
@lcnr perf looks relatively the same after the latest push (albeit a tiny bit worse - though this might be noise). Interestingly when I build the `packed-simd` benchmark locally, this PR's build is almost always faster (though maybe not statistically significantly so).

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 1]
```diff
@@ -67,6 +67,37 @@ where
...
+    let impl2_self = tcx.type_of(impl2_def_id);
+    let impl1_ref = tcx.impl_trait_ref(impl1_def_id);
+    let impl2_ref = tcx.impl_trait_ref(impl2_def_id);
+
+    // Check if any of the input types definitely do not unify.
+    if impl1_ref
+        .iter()
+        .flat_map(|tref| tref.substs.types())
+        .zip(impl2_ref.iter().flat_map(|tref| tref.substs.types()))
+        .chain(iter::once((impl1_self, impl2_self)))
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:83]: that isn't required, is it?

The impl_self should be `tref.substs[0]`, so you're currently checking that one twice

**Pull request merged**